include/ginkgo/core/base/precision_dispatch.hpp

```diff
@@ -302,13 +302,24 @@ void mixed_precision_dispatch_real_complex(Function fn, const LinOp *in,
     if (complex_to_real) {
         mixed_precision_dispatch<to_complex<ValueType>>(
             [&fn](auto dense_in, auto dense_out) {
-                using Dense = matrix::Dense<ValueType>;
+                using DenseIn = typename std::decay_t<decltype(*dense_in)>;
+                using DenseOut = typename std::decay_t<decltype(*dense_out)>;
+                using DenseDummy = matrix::Dense<ValueType>;
```

Comment:
Maybe add a comment that you need `DenseDummy`, so it compiles properly (at least I think that is the reason).
Additionally, think about using SFINAE on this whole function `mixed_precision_dispatch_real_complex` to ensure that `ValueType` is complex, or not. IMO, this would make it more readable and avoid the work-around of the `dynamic_cast` (and potentially the `DenseDummy`).

Reply:
Good point. For the older implementation, this direct cast solution is preferable, but for this more complex setting, SFINAE/enable_if seems preferable